homeassistant/components/mailgun/__init__.py

```diff
@@ -0,0 +1,104 @@
+"""
+Support for Mailgun.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/mailgun/
+"""
+from urllib.parse import urlparse, parse_qsl
+
+import voluptuous as vol
+from ipaddress import ip_address
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant import config_entries
+from homeassistant.const import CONF_API_KEY, CONF_DOMAIN, CONF_WEBHOOK_ID
+from homeassistant.util.network import is_local
+
+DOMAIN = 'mailgun'
+API_PATH = '/api/{}'.format(DOMAIN)
+DEPENDENCIES = ['webhook']
+MESSAGE_RECEIVED = '{}_message_received'.format(DOMAIN)
+CONF_SANDBOX = 'sandbox'
+DEFAULT_SANDBOX = False
+
+CONFIG_SCHEMA = vol.Schema({
+    vol.Optional(DOMAIN): vol.Schema({
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Required(CONF_DOMAIN): cv.string,
+        vol.Optional(CONF_SANDBOX, default=DEFAULT_SANDBOX): cv.boolean,
+        vol.Optional(CONF_WEBHOOK_ID): cv.string,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the Mailgun component."""
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN] = config[DOMAIN]
+    return True
+
+
+async def handle_webhook(hass, webhook_id, request):
+    """Handle incoming webhook with Mailgun inbound messages."""
+    data = await request.post()
+
+    if isinstance(data, dict):
+        data[CONF_WEBHOOK_ID] = webhook_id
+    hass.bus.async_fire(MESSAGE_RECEIVED, data)
+
+
+async def async_setup_entry(hass, entry):
+    """Configure based on config entry."""
+    hass.components.webhook.async_register(
+        entry.data[CONF_WEBHOOK_ID], handle_webhook)
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
+    return True
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class ConfigFlow(config_entries.ConfigFlow):
+    """Handle a Mailgun config flow."""
+
+    async def async_step_user(self, user_input=None):
+        """Handle a user initiated set up flow."""
+        if self._async_current_entries():
```

Comment:
A lot of this code is directly copied from the IFTTT setup. I would be willing to create a superclass for use for both mailgun and ifttt config flow but I'm not sure if that's the best way of handling this duplication.

Is there any way to share some of the strings? Any value in doing so?

Reply:
Good call on generalizing! 

We should follow the same strategy as the generalized discovery flow: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_entry_flow.py#L7

Strings will still be done on a per component basis, which is good as for webhooks it can be all over the place.